### PR TITLE
Patchy particle pressure

### LINKED ISSUE #43: Validate expansive SDF.
## Description

Add tests that exercise the expansive SDF: https://github.com/glotzerlab/hoomd-blue/pull/1391
* Concave, interlocking particles.
* A non-monotonic potential (i.e. shifted square well).

## Proposed solution

Follow the existing test suites and provide NVT/NPT consistency checks.

## Additional context

The GPU code path for user potentials is already well tested. These tests need only run on the CPU to verify that SDF works properly.

## Developer

@tcmoore3, can you implement this before the HOOMD v4 release?

### PR DESCRIPTION
**Pull request by tcmoore3**
## Description

Validate pressure computation for discontinuous pair potentials simulated with HPMC.

I also changed the comparison plots to show the distributions of sampled pressures and densities.

## Motivation and context

The pressure pressure computation for systems with discontinuous pair potentials in HPMC is currently not tested for validity but should be. 

I will add another validation test for the hard sphere + square well + square shoulder system we discussed, but I wanted to get this PR opened because the code will not change very much.

Resolves #43. 

## How has this been tested?

We discovered a conceptual error in the implementation of the SDF calculation for systems with pair potentials. We implemented a fix at https://github.com/glotzerlab/hoomd-blue/commit/8d7cb2c7f3c6abf3326aeea6ca26af5a062057f5. That commit is the version of HOOMD-blue I used to generate the results below.

I ran NVT simulations of 4,096 particles interacting either through a hard sphere + square well potential, or through a Kern–Frenkel potential, i.e., a hard sphere + square well model with an angular component to the pair potential. I ran NVT simulations, measured the pressure for the given density and temperature, and then ran NPT simulations at the corresponding pressure. The agreement between the pressure measured in NVT simulations and the density measured at the corresponding NPT simulations is shown in the figures below. These results are from 16 replicates at each statepoint. \chi = 1.0 corresponds to the hard sphere + square well state points, and \chi = 0.7 is the Kern–Frenkel model.

image
image
image
image
image

## Checklist:

- [x] I have reviewed the [**Contributor Guidelines**](https://github.com/glotzerlab/hoomd-validation/blob/trunk/CONTRIBUTING.md).
- [x] I agree with the terms of the [**HOOMD-blue Contributor Agreement**](https://github.com/glotzerlab/hoomd-validation/blob/trunk/ContributorAgreement.md).
- [x] My name is on the list of contributors (`CONTRIBUTORS.md`) in the pull request source branch.

### CODE REVIEW

**Comment by joaander** [comment]
With short simulations, I tested:
```
sigma = 1.0
lambda_ = 1.5
epsilon = 1/3.0
```
with patch code:
```
    evaluate_square_well = f'''float rsq = dot(r_ij, r_ij);
                               if (rsq < {(lambda_ * sigma/2)**2}f)
                                   return param_array[0];
                               else if (rsq < {(lambda_ * sigma)**2}f)
                                   return param_array[1];
                               else
                                   return 0.0f;
                            '''
    square_well = hoomd.hpmc.pair.user.CPPPotential(r_cut=lambda_ * sigma,
                                                    code=evaluate_square_well,
                                                    param_array=[-epsilon, epsilon])
```
at a variety of pressures (2-10) and densities (around ~0.7).

File: hoomd_validation/config_parser.py [hunk 1]
```diff
@@ -34,7 +34,7 @@ def __init__(self, config_file_path=DEFAULT_CONFIG_PATH):
             config.get("max_cores_submission", 16))
         self["max_gpus_submission"] = int(config.get("max_gpus_submission", 1))
         self["max_walltime"] = int(config.get("max_walltime", 24))
```
Comment (joaander) [inline hoomd_validation/config_parser.py]:
```suggestion
        self["max_walltime"] = float(config.get("max_walltime", 24))
```
For consistency with the below change.

File: hoomd_validation/init.py [hunk 2]
```diff
@@ -14,14 +14,16 @@
...
-    lj_union,
-    hard_disk,
-    hard_sphere,
-    simple_polygon,
+#    alj_2d,
+#    lj_fluid,
+#    lj_union,
+#    hard_disk,
+#    hard_sphere,
+#    simple_polygon,
```
Comment (joaander) [inline hoomd_validation/init.py]:
```suggestion
   alj_2d,
   lj_fluid,
   lj_union,
   hard_disk,
   hard_sphere,
   simple_polygon,
```

File: hoomd_validation/init.py [hunk 3]
```diff
@@ -14,14 +14,16 @@
...
+    patchy_particle_pressure
```
Comment (joaander) [inline hoomd_validation/init.py]:
```suggestion
    patchy_particle_pressure,
```

File: hoomd_validation/util.py [hunk 4]
```diff
@@ -28,7 +28,7 @@ def get_job_filename(sim_mode, device, name, type):
     return f"{sim_mode}_{suffix}_{name}.{type}"
 
 
-def run_up_to_walltime(sim, end_step, steps, walltime_stop):
+def run_up_to_walltime(sim, end_step, steps, walltime_stop, minutes_per_run=None):
```
Comment (joaander) [inline hoomd_validation/util.py]:
```suggestion
def run_up_to_walltime(sim, end_step, steps, walltime_stop):
```
unused.

File: hoomd_validation/util.py [hunk 5]
```diff
@@ -168,30 +168,49 @@ def plot_distribution(ax, data, xlabel, expected=None, bins=100):
...
+        bin_centers = 0.5 * (bin_edges[:-1] + bin_edges[1:])
 
         if numpy.all(data_arr == data_arr[0]):
             histogram[:] = 0
 
         max_density_histogram = max(max_density_histogram, numpy.max(histogram))
 
-        ax.plot(bin_edges[:-1], histogram, label=mode)
+        if plot_rotated:
+            XX, YY = histogram, bin_centers #bin_edges[:-1]
```
Comment (joaander) [inline hoomd_validation/util.py]: Remove the need for the variables by placing the call to `plot` inside the conditional.

Alternately, replace meaningless XX and YY variable names with something meaningful.

File: hoomd_validation/util.py [hunk 6]
```diff
@@ -168,30 +168,49 @@ def plot_distribution(ax, data, xlabel, expected=None, bins=100):
...
+            ax.set_ylabel(xlabel)
```
Comment (joaander) [inline hoomd_validation/util.py]: The `xlabel` is no longer the x label. Change the name of the argument appropriately.

File: hoomd_validation/util.py [hunk 7]
```diff
@@ -168,30 +168,49 @@ def plot_distribution(ax, data, xlabel, expected=None, bins=100):
...
+        ax.plot(XX, YY, label=mode)
 
-    ax.set_xlabel(xlabel)
-    ax.set_ylabel('probability density')
 
     if callable(expected):
-        ax.plot(bin_edges[:-1],
-                expected(bin_edges[:-1]),
+        if plot_rotated:
+            #Y, X = bin_edges[:-1], expected(bin_edges[:-1])
```
Comment (joaander) [inline hoomd_validation/util.py]: Remove commented code.

File: hoomd_validation/util.py [hunk 8]
```diff
@@ -168,30 +168,49 @@ def plot_distribution(ax, data, xlabel, expected=None, bins=100):
...
 
         max_density_histogram = max(max_density_histogram, numpy.max(histogram))
 
-        ax.plot(bin_edges[:-1], histogram, label=mode)
+        if plot_rotated:
+            XX, YY = histogram, bin_centers #bin_edges[:-1]
+            ax.set_ylabel(xlabel)
+            ax.set_xlabel('probability density')
+        else:
+            XX, YY = bin_edges[:-1], histogram
```
Comment (joaander) [inline hoomd_validation/util.py]: The unrotated plot should also use `bin_centers`.

File: hoomd_validation/util.py [hunk 9]
```diff
@@ -168,30 +168,49 @@ def plot_distribution(ax, data, xlabel, expected=None, bins=100):
...
-    ax.set_ylabel('probability density')
 
     if callable(expected):
-        ax.plot(bin_edges[:-1],
-                expected(bin_edges[:-1]),
+        if plot_rotated:
+            #Y, X = bin_edges[:-1], expected(bin_edges[:-1])
+            Y, X = bin_centers, expected(bin_centers)
+        else:
+            X, Y = bin_edges[:-1], expected(bin_edges[:-1])
```
Comment (joaander) [inline hoomd_validation/util.py]: The unrotated plot should also use `bin_centers`.

File: hoomd_validation/util.py [hunk 10]
```diff
@@ -168,30 +168,49 @@ def plot_distribution(ax, data, xlabel, expected=None, bins=100):
...
 
-    ax.set_xlabel(xlabel)
-    ax.set_ylabel('probability density')
 
     if callable(expected):
-        ax.plot(bin_edges[:-1],
-                expected(bin_edges[:-1]),
+        if plot_rotated:
+            #Y, X = bin_edges[:-1], expected(bin_edges[:-1])
+            Y, X = bin_centers, expected(bin_centers)
```
Comment (joaander) [inline hoomd_validation/util.py]: Remove the need for the variables by placing the call to `plot` inside the conditional.

Alternately, name these variables with a PEP8 compliant name.

File: hoomd_validation/util.py [hunk 11]
```diff
@@ -299,3 +318,46 @@ def plot_timeseries(ax,
 def _sort_sim_modes(sim_modes):
     """Sort simulation modes for comparison."""
     sim_modes.sort(key=lambda x: ('nvt' in x or 'nec' in x, 'md' in x, x))
+
+
+def _single_patch_kern_frenkel_code(delta_rad, lambda_, sigma, kT):
```
Comment (joaander) [inline hoomd_validation/util.py]: Move this function to `patchy_particle_pressure.py`.

File: hoomd_validation/patchy_particle_pressure.py [hunk 12]
```diff
@@ -0,0 +1,725 @@
...
+    directives = dict(walltime=CONFIG["max_walltime"],
+                      executable=CONFIG["executable"],
+                      nranks=util.total_ranks_function(ranks_per_partition))
+
+    if device_name == 'gpu':
+        directives['ngpu'] = directives['nranks']
+
+    @Project.pre.after(patchy_particle_pressure_create_initial_state)
+    @Project.post.isfile(f'{mode}_{device_name}_complete')
+    #@Project.post(lambda j: j.isfile(f'{mode}_{device_name}_complete') or j.sp.pressure <= 0)
```
Comment (joaander) [inline hoomd_validation/patchy_particle_pressure.py]:
```suggestion
```

**Review by joaander — CHANGES_REQUESTED** [review]
I see you have GPU jobs definitions in the workflow. Do you plan to validate that these work properly before merging?

**Comment by tcmoore3** [comment]
Results for the hard sphere + square well + repulsive shoulder potential. It shows similar quantitative agreement NPT and NVT simulations as the Kern–Frenkel and hard sphere + square well systems. 
image

File: hoomd_validation/patchy_particle_pressure.py [hunk 13]
```diff
@@ -0,0 +1,822 @@
...
+        device = device_cls(
+            communicator=communicator,
+            message_filename=job.fn(f'{mode}_{device_name}.log'))
+
+        globals().get(f'run_{mode}_sim')(
+            job, device, complete_filename=f'{mode}_{device_name}_complete')
+
+        if communicator.rank == 0:
+            print(f'completed patchy_particle_pressure_{mode}_{device_name} '
+                  f'{job} in {communicator.walltime} s')
```
Comment (joaander) [inline hoomd_validation/patchy_particle_pressure.py]:
```suggestion
            print(f'completed patchy_particle_pressure_{mode}_{device_name}: {job}')
```
See #67.

File: hoomd_validation/patchy_particle_pressure.py [hunk 14]
```diff
@@ -0,0 +1,822 @@
...
+    hoomd.write.GSD.write(
+        state=sim.state,
+        filename=job.fn("patchy_particle_pressure_initial_state.gsd"),
+        mode='wb',
+        logger=trajectory_logger,
+    )
+
+    if communicator.rank == 0:
+        print(f'completed patchy_particle_pressure_create_initial_state: '
+              f'{job} in {communicator.walltime} s')
```
Comment (joaander) [inline hoomd_validation/patchy_particle_pressure.py]:
```suggestion
        print(f'completed patchy_particle_pressure_create_initial_state: {job}')
```
See #67.

**Review by joaander — CHANGES_REQUESTED** [review]
More fixes to prevent deadlock.

**Comment by tcmoore3** [comment]
GPU results from 8 replicates at each of 3 statepoints, 1 KF, 1 HSSW, 1 HSSW with a repulsive shoulder. Note that I took the average pressure from the CPU NVT runs, so the deviation from the expected pressure in the 'nvt_gpu' runs are nonzero.
image
image
image

File: hoomd_validation/patchy_particle_pressure.py [hunk 15]
```diff
@@ -0,0 +1,820 @@
...
+        # hard sphere + square well statepoints, from 10.1063/1.3054361
+        # pressure from NVT simulations, NOT from values in paper
+        # (0.5714, 0.8, -0.2692376274894095, 1.0, 1.5, 1.0),
+        # (1.0, 0.8, 2.2766339608381325, 1.0, 1.5, 1.0),
+        (3.0, 0.8, 4.837436833423719, 1.0, 1.5, 1.0),
+        # hard sphere + square well + repulsive shoulder statepoints.
+        # temperatures/densities from initial tests, pressures from NVT
+        # simulations.
+        (3.0, 0.7, 4.00804, 1.0, 1.0, -1.0),
+        # (1.0, 0.7, -5.0, 1.0, 1.0, -1.0),
```
Comment (joaander) [inline hoomd_validation/patchy_particle_pressure.py]:
```suggestion
        (1.0, 0.95, 10.208625410213362, 0.7, 1.5, 1.0),
        # hard sphere + square well statepoints, from 10.1063/1.3054361
        # pressure from NVT simulations, NOT from values in paper
        (3.0, 0.8, 4.837436833423719, 1.0, 1.5, 1.0),
        # hard sphere + square well + repulsive shoulder statepoints.
        # temperatures/densities from initial tests, pressures from NVT
        # simulations.
        (3.0, 0.7, 4.00804, 1.0, 1.0, -1.0),
```
Remove unused state points.

**Review by joaander — CHANGES_REQUESTED** [review]
Thanks! There is one minor change to make, then this is ready to merge.

**Review by joaander — APPROVED** [review]
Thanks!

**Pull request merged**